### ISSUE TITLE
New DB design

### ISSUE THREAD

**Issue by ChihChengLiang** [thehubbleproject/hubble-commander#103]
Writing down our discussion last week

## TL;dr

- The schema is designed for key-value database, but it also works with relational DB.
- Old nodes not pruned. Pros: don't worry about rollback. Cons: tree size bloating
- "Get" a leaf needs `tree_depth` reads, 
- "Update"  a leaf needs  `tree_depth` reads and `tree_depth` writes

### Public Key

key
```
"pubkey" + pubkeyID
```

value
```
publickeyBytes
```

### State

key
```
"state" + stateID
```

value
```
serialize(
    pubkeyID,
    tokenID,
    balance,
    nonce
)
```

#### Nodes

We have to build merkle tree for state and pubkeys. 

key
```
"tree" + "state" + leafHash
```
value

```
child_left + child_right
```

#### Root

key
```
"stateTreeRoot"
```
value

```
stateTreeRoot
```

### Batch
key
```
"batch" + batchID
```
value
```
serialize(
    PostStateRoot,
    Committer,
    Commitments[],
    BatchType
)
```

### Token Registry?

Should we do tokenRegistry?

- Yes, if we want to query address
- No, the source of truth is on chain

### Params

Should we do params?

### Relayer?

What is relayer?

### Tx

Should we do Tx with database? Or they can just stay in memory?

## API

TODO caching

### Abstract Tree

Note: Old nodes not pruned. Pros: don't worry about rollback. Cons: tree size bloating

Reads: num_leaf * 32
Writes: num_leaf * 32

```
def get(db, root, path) -> Witness:
    key = root
    witness = []
    for depth, bit in enumerate(path):
        if key == defaultHash[depth]:
            return witness + defaultHash[depth+1:]
        child_left, child_right = db.get(key)
        if bit == 0:
            key = child_left
            witness.append(child_right)
        else:
            key = child_right
            witness.append(child_left)
    return witness
```

Reads: num_leaf * 32
Writes: num_leaf * 32

```
def update(db, root, path, leaf) -> Root:
    witness = get(db, root, path)
    node = leaf
    for height, bit in enumerate(reverse(path)):
        if bit == 0:
            key = hash(witness[height], node)
        else:
            key = hash(node, witness[height])
        db.put(key, node)
        node = key
    return node
```

### Public key

TODO: handle bulk case

```
def GetPublicKey(pubkeyID):
    pubkey = db.pubkey.get(pubkeyID)
    root = db.pubkeyRoot.get()
    get(db.pubkeyTree, root, getPath(pubkeyID))
```

```
def UpdatePublicKey(pubkeyID, key):
    db.pubkey.put(pubkeyID, key)
    root = update(db.pubkeyTree, root, getPath(pubkeyID), hash(key))
    db.pubkeyRoot.put(root)
```

### State

Almost the same as public key

```
def GetState(stateID):
def UpdateState(stateID, state):
```

Could have this to check tx nonce
```
def GetStateNoWiteness(stateID):
```

### Batch

```
GetBatch(batchID)
GetLatestBatch()
```

### Auxiliary

GetStateCount()?

**Comment by vaibhavchellani** [comment]
A majority of this has been implemented in #111
